### PR TITLE
stereotool: init at 10.21

### PR DESCRIPTION
**Pull request by RudiOnTheAir**
## Description

Stereo Tool is a software-based audio processor which offers outstanding audio quality and comes with many unique features. It is used by over 3000 FM stations ranging from small local stations to 50-100 kW stations and nation-wide networks with dozens of transmitters, thousands of streaming stations and many DAB+, HD, AM and TV stations. It can be used for both live and file based processing.

Beside performing traditional audio processing, Stereo Tool can also repair many common audio problems on the fly, and optimize outgoing audio for better FM reception. A full list of features can be found below on this page.

[https://www.thimeo.com/stereo-tool/](url)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] x686-linux
  - [x] aarch64-linux
  - [x] aarch32-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by tobiasBora** [comment]
Congrats for your first PR! Sadly, I think you have not squashed correctly your commits as we [discussed here](https://discourse.nixos.org/t/how-to-find-needed-librarys-for-closed-source-bin-applications/39118/86): if you go to the [commit tab](https://github.com/NixOS/nixpkgs/pull/288992/commits), you will see that a single commit is present instead of two (one for the maintainer, one for the package)… and worse, you also squashed some commits made before yours by others in systemd (explaining the merge issue)… which must definitely be fixed. As I know you are not used to git, the simpler is maybe to:

- first, backup the two files you want to push
```
$ cp maintainers/maintainer-list.nix .
$ cp pkgs/by-name/st/stereotool/package.nix stereotool-package.nix
$ cp pkgs/by-name/st/stereotool/stereo-tool-icon.png .
```
- then move your branch one commit above (you lose the systemd changes, but they will be included back after the PR is merged so no worries)
```
$ git checkout -B add_stereotool HEAD^
```
- check with:
```
$ git log
```
that you see none of your commits.
- Then, restore the maintainer-list file first:
```
$ cp . maintainers/maintainer-list.nix
```
With `git diff` (that prints what will be commited next), make sure that only your name is added to the file and no other modification is made. If so, commit with:
```
$ git commit -am "maintainers: add RudiOnTheAir"
```
(you should now see this commit listed in `git log`)
- Finally, copy the icon and package back:
```
$ cp stereotool-package.nix pkgs/by-name/st/stereotool/package.nix
$ cp stereo-tool-icon.png pkgs/by-name/st/stereotool/
$ git add pkgs/by-name/st/stereotool/package.nix pkgs/by-name/st/stereotool/stereo-tool-icon.png
```
and do again a `git diff` to check that you are only committing these two files. Then, do the commit:
```
$ git commit -am "stereotool: init at 10.21"
```
Now, `git log` should show you two commits of yours (you can inspect them individually if you want with `git show IDOFCOMMIT`). Your can then push the changes as before, using:
```
$ git push --force myfork HEAD
```

**Comment by tobiasBora** [comment]
Btw, once you followed the above instructions, if later you want to change stuff in your derivation, instead of doing a commit + squash, just commit with:
```
$ git commit --amend --no-edit
```
and it will modify the last commit directly (then, force push)

**Comment by RudiOnTheAir** [comment]
> ... followed the above instructions

Done that. Looks better. And i hope that the future maintanance of this package makes u not so much work (for you)

**Comment by RudiOnTheAir** [comment]
There is one check left... and two merge conflicts

**Comment by tobiasBora** [comment]
Much better yeah. Just, I just realized, can you remove the space as I mentioned in the thread, like change:
```
{ lib
  , stdenv
  , fetchurl
  , autoPatchelfHook
  , libX11
  , libXpm
  , alsa-lib
  , bzip2
  , zlib
  , kdialog
  , libgcc
  , makeWrapper
  , copyDesktopItems
  , makeDesktopItem
}:
```
into
```
{ lib
, stdenv
, fetchurl
, autoPatchelfHook
, libX11
, libXpm
, alsa-lib
, bzip2
, zlib
, kdialog
, libgcc
, makeWrapper
, copyDesktopItems
, makeDesktopItem
}:
```
This formatting is quite important to maintain a uniform style across packages (and so far it worked fairly well, only 2 packages in the whole nixpkgs use this bad style). Soonish, I think a new formatting standard will be applied (maybe even automatically) following a new RFC, but so far we need to do it manually and use this style (make sure to do your commit with `git commit -a --amend --no-edit`)

But I'm not sure to understand why fborg complains… https://gist.github.com/GrahamcOfBorg/9c87792a8a3d2c91903c317e717d6cc3, might be related? https://github.com/NixOS/ofborg/issues/518

**Review by tobiasBora — APPROVED** [review]
Seems good to me.

File: pkgs/by-name/st/stereotool/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/st/stereotool/package.nix:22]:
```suggestion

  srcs = let
    versionNoPoint = lib.replaceStrings [ "." ] [ "" ] version;
  in {
```

File: pkgs/by-name/st/stereotool/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/st/stereotool/package.nix:101]:
```suggestion
```
we don't need that

File: pkgs/by-name/st/stereotool/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/st/stereotool/package.nix:124]:
```suggestion
      name = "stereotool-alsa";
      desktopName = "Stereotool-Alsa";
      exec = "stereo_tool_gui";
      icon = "stereo-tool-icon";
      comment = "Broadcast Audio Processing";
      categories = [ "AudioVideo" "Audio" "AudioVideoEditing" ];
    })
    (makeDesktopItem {
      name = "stereotool-jack";
```

File: pkgs/by-name/st/stereotool/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/st/stereotool/package.nix:139]: That should be already in stdenv. If this is here to fix the darwin built, we should gccStdenv instead of stdenv.

File: pkgs/by-name/st/stereotool/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/st/stereotool/package.nix:150]: Can we fetch that instead?

File: pkgs/by-name/st/stereotool/package.nix
Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:124]: Out of curiosity, why is this better?

File: pkgs/by-name/st/stereotool/package.nix
Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:150]: The author of this PR cannot find any icon online, he actually made this icon himself from a cut of the online banner, not sure if there is a better solution (and the image is quite small).

File: pkgs/by-name/st/stereotool/package.nix
Comment (kirillrdy) [reply pkgs/by-name/st/stereotool/package.nix:124]: https://github.com/NixOS/nixpkgs/issues/277994

File: pkgs/by-name/st/stereotool/package.nix
Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:150]: I could ask the maintainer for storing the icon on his storage... If that is the better solution??

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:150]: If you ask that would be great

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:150]: But make it clear that the link must be stable with fixed content, i.e. same hash.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:149]: I've asked hin. He said:
_What I can also do is upload it alongside the binaries, that way each version can theoretically have a different logo. Would that make sense, or not at all? If we ever want to change it, it wouldn't change for older versions if we do it that way (and if you're hashing it that might be a good idea)._
I'm waiting for the URL like this https://download.thimeo.com/stereo_tool_logo_10.png

File: pkgs/by-name/st/stereotool/package.nix
Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:138]: Yes, that's what Tobias meant, but it doesn't seem to work without it (libgcc)

[https://discourse.nixos.org/t/how-to-find-needed-librarys-for-closed-source-bin-applications/39118/57?u=rudiontheair](url)

File: pkgs/by-name/st/stereotool/package.nix
Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:149]: Yeah, it's exactly what we want.

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:149]: @SuperSandro2000 does rudi need to squash the commits or the commiter can do it? Not sure if they can when a new maintainer is added.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:149]: I'm waiting for getting the link...

Comment (SuperSandro2000) [reply pkgs/by-name/st/stereotool/package.nix:149]: Yes, please squash the review commits to have a clean history.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:149]: Honesly, i'm not sure if it is ok now?

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:149]: You should squash it at the very end once you add the image.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:149]: OK, still waiting that the image is available at the download location. Don't want to be impatient now.. Then the download must also be entered in package.nix. Probably similar to the binaries...?!

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: OK, to get it running, i have added the icon to my webserver, later it would be simple to switch over to the maintainers url...
OK, but there it something wrong and it has to do with the changes related to the icon download part... And i don't believe i have to change that ; to )..?!

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: Yes, i know, but a thought that was such a simple change... Sadly i can dive deeper in this tomorrow. Did u already changed it? I can see only a diff in the copy icon line..??

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix]: No, I don't have the right to change it myself. I can only propose a change (you would need to accept it), and since anyway the indentation needs to be fixed, I guess it is simpler to just do it yourself. No worries, there is nothing urgent ;-)

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: What do you think about nixpkgs-fmt?

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix]: Yeah, it is definitely giving very good results, you can definitely use it (but double check notably for the inline code), see also the web version here https://nix-community.github.io/nixpkgs-fmt/ The community is working on a new one with new rules from an RFC, but so far it has not been deployed at as far as I know, so use nixpkgs-fmt.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: As u know, i'm using kate to edit, and it marks parenthesis yellow when detecting two which belongs together. The only one if found without yellow color ist on line 30.
` ] ++ (`
And i don't think the parenthesis is missing in line 107
`  }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}");
`
So where is it missing...?

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix]: Yes, exactly, the parenthesis on line 30 should be closed on line 107, so that you get:
```
] ++ (
  {
    … 
  }.${stdenv.hostPlatform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}"));
```

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: .. and one day i hopefully understand why it was missing on that position and not on any other... I mean i had error messages that not point direct to error... ?! I had ran the fmt tool and would commit then...

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix]: Ahaha. The reason is quite simple. We are building a list of terms like:
```
[ 42 43 44 ]
```
(in practice, `42`, `43` and `44` are replaced with the derivation pointing to the downloaded files, but it is not needed here to understand the syntax). What we do right now, first is to cut this list in two parts, one independent of the architecture (`42` in our example), and two that differs depending on the architecture (`43` and `44`). We can combine two lists into a single list via `++`, like in:
```
[ 42 ] ++ [ 43 44 ]
```
Then, it costs nothing to add a parenthesis, like `1 + 1` is the same as `1 + (1)`, so similarly, we can equivalently write the above as:
```
[ 42 ] ++ ( [ 43 44 ] )
```
I guess you would never write `1 + (1` right? It is the precise reason why the parenthesis above must be added a the very end of the expression. Not too complicated right?

You might wonder here, why we bother adding a parenthesis. The reason is that we will complexify the right part, as it is not always `[ 43 44 ]`: if the `system` is different, it might be like `[ 143 144 ]`. The trick here is to obtain this list via a "dictionary" if you write:
```
let mydict = {
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
};
```
then `mydict."x86_64-linux"` will be equal to `[43 44]`, hence `mydict.${stdenv.hostPlateform.system}` will be equal to `[ 43 44 ]` if the system is `x86_64-linux` and `[ 143 144 ]` if the system is `aarch64-linux`. But writing:
```
let mydict = {
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
}; in
mydict.${stdenv.hostPlateform.system}
```
is a bit long, so we can simplify it by writting everything in a single line like in:
```
# This outputs either [ 43 44 ] or [ 143 144 ]:
{
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
}.${stdenv.hostPlateform.system}
```
The only problem now is that it might fail if the plateform is not listed (like what happens if `system` is `aarch64-darwin`?). The syntax in nix is to add `or (throw "Unsupported system: ${stdenv.hostPlatform.system}")` to raise an error if the lookup to the dictionary is empty. Hence, we get:
```
# This outputs either [ 43 44 ] or [ 143 144 ]:
{
  "x86_64-linux" = [ 43 44];
  "aarch64-linux" = [ 143 144];
}.${stdenv.hostPlateform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}")
```
Since this expression still evaluates to `[ 43 44 ]` or alike, we can replace the `[ 43 44 ]` in `[ 42 ] ++ ( [ 43 44 ] )` with this expression, giving us:
```
[ 42 ] ++ (
  {
    "x86_64-linux" = [ 43 44];
    "aarch64-linux" = [ 143 144];
  }.${stdenv.hostPlateform.system} or (throw "Unsupported system: ${stdenv.hostPlatform.system}"))
```
(see, we have two `))` at the end, as expected). The very first and very last parenthesis are not superfluous anymore, as now the inner expression is more complex (similarly, we cannot remove the parenthesis in `2 * (2 + 3)` as it is not the same as `2 * 2 + 3`). Is it a bit clearer?

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: Yes, I understand that so far. It's difficult for a beginner to see what's missing and where when it's so far apart.

That being said, after the commit, it's still being complained about. Let's see why...

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix]: @RudiOnTheAir The latest version seems good to me (except for the indentation of `unpackPhase`, which is properly set if you use https://nix-community.github.io/nixpkgs-fmt/, and the commits where you don't anymore have a single commit mentioning `maintainers: add XXX`), and of course the URL that must be updated once it appears officially online.

The reason ofborg complains is actually more subtle, i is because of a merge conflict (apparently tests are not re-run when a merge conflict occurs). Indeed, you were quite unlucky, as someone added their name in the maintainer list precisely where you were supposed to be added, so now Git does not know if they need to replace this name with yours, add yours before or above etc. So you will have to fix this merge conflict, basically by restarting from the latest `master` branch. Since anyway you need to create a commit `maintainers: add XXX`, and the commit history starts to be a bit ugly, the simplest to solve this is certainly to restart from a clean `git fetch` of the master branch. But it might be worth doing that only once you have the official URL of the picture online, it will save you some squash later.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix]: > (except for the indentation of `unpackPhase`, which is properly set if you use https://nix-community.github.io/nixpkgs-fmt/,

I don't understand. I had used that tool direct before committing and it has made changes. I had installed that tool as a package.. And the lines around "unpackPhase".. i havn't touched them after after i had corrected it after your discourse post. (the quote ruined it here, i know) But that seems to do the fmt tool, too...?!

> should be:
 unpackPhase = ''
    for srcFile in $srcs; do
      cp $srcFile $(stripHash $srcFile)
    done
  '';

Honestly i don't how to go futher. I was lucky to work with the "github-desktop". I can imagine that i could use the cmdline tools more one day. But some steps are a bit easier and more comfortable with it to dive into git. I had compiled my radio software for many years, had a private debian repo with selfmade packages, so i thought nix is the new challenge. But that makes not much sence when i have to ask every day.. What do you mean? Maybe the discourse would be the better place to discuss...;)

**Comment by RudiOnTheAir** [comment]
> Otherwise the indentation looks better now.

Don't know if i should be lucky about that.

**Comment by tobiasBora** [comment]
I see that you tried to push a merge… but it does not change anything (your commit explicitly says that you should remove these authors, so a merge will not change that. I'm afraid you will need to restart again the procedure, by cleanly adding your name in the list without copy/pasting the list of maintainers as I recommended in the other thread… and by triple checking if the commit contains what it is supposed to contain before pushing.

File: pkgs/by-name/st/stereotool/package.nix
Comment (kirillrdy) [inline pkgs/by-name/st/stereotool/package.nix:162]: mainProgram should be set

**Review by kirillrdy — COMMENTED** [review]
```
error: builder for '/nix/store/rkp7bihdqixzvxk8gb1ssahbzzv35ygi-stereo-tool-icon.png.drv' failed with exit code 1;
       last 7 log lines:
       >
       > trying https://download.thimeo.com/stereo_tool_logo_1021.png
       >   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
       >                                  Dload  Upload   Total   Spent    Left  Speed
       >   0   196    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
       > curl: (22) The requested URL returned error: 404
       > error: cannot download stereo-tool-icon.png from any mirror
       For full logs, run 'nix log /nix/store/rkp7bihdqixzvxk8gb1ssahbzzv35ygi-stereo-tool-icon.png.drv'.
```

**Comment by RudiOnTheAir** [comment]
> trying https://download.thimeo.com/stereo_tool_logo_1021.png
Typo, had to be icon, not logo.. Better don't ask... ;( Changed

File: pkgs/by-name/st/stereotool/package.nix
Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:162]: How to handle in this case two two programs stereotool-alsa and stereotool-jack? Choose one ot them?
> description = "Stereo Tool is a software-based audio processor which offers outstanding audio quality and comes with many unique features.";
    license = licenses.unfree;
    # mainProgram = "stereotool-alsa";
    platforms = [ "aarch64-linux" "aarch32-linux" "x86_64-linux" "i686-linux" ];

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:162]: Yeah, just put the alsa. Anyway, it seems to refer to the same binary for now ^^'

Comment (tobiasBora) [reply pkgs/by-name/st/stereotool/package.nix:162]: But you want `mainProgram = "stereo_tool_gui"` since it is the name of the corresponding binary.

Comment (RudiOnTheAir) [reply pkgs/by-name/st/stereotool/package.nix:162]: OK, later. Now I had to make my Radioshow ready for transmission an 1800 today. Running Rivendell and stereotool... Rivendell. The next big thing on nix?...:o)

**Review by kirillrdy — APPROVED** [review]
i only tried x86_64-linux. but builds and runs

**Review by tobiasBora — COMMENTED** [review]
Good for me, thanks!

**Pull request merged**
